lib/conrad/emitter_queue.rb

```diff
@@ -0,0 +1,58 @@
+module Conrad
+  # Centralized event emission queue across threads
+  class EmitterQueue
+    include Singleton
+
+    # Boolean that determines whether events will be emitted inline or in a
+    #   background thread
+    attr_reader :background
+
+    # Logger object used for sending log events
+    attr_accessor :logger
+
+    def initialize
+      @thread = nil
+      @queue = Queue.new
+      @logger ||= Logger.new(STDOUT)
+    end
+
+    # bakground setter. Will start/stop the background thread
+    # @param value [Boolean] assigns whether events should be processed inline
+    #   or in a separate thread.
+    def background=(value)
+      @background = value
+      value ? start_thread : @thread && @thread.join
+    end
+
+    # Enqueues a block
+    # @yield block to execute. Will either run inline or separately depending on
+    #   whether the queue is backgrounded
+    def enqueue
+      @queue.push -> { yield }
+
+      # if it's backgounded we can break out of here, as the background
+      #   queue will pick it up. otherwise, we need to explicitly process it
+      emit! unless @background
+    end
+
+    private
+
+    def start_thread
+      @thread ||= Thread.new do
+        Thread.current.abort_on_exception = true
+        loop do
+          emit!
+          break unless Conrad::EmitterQueue.instance.background
```

Comment:
I feel like I'm not understanding the purpose of this `break`. Would you mind elaborating on it?

Reply:
You had previously mentioned that in order to get garbage collected we need to `join` the dequeueing thread. Did i understand that correctly? If we were to simply join it, we would be joining a `loop do`. This way once someone decides to unbackground the processing, we will finish whatever work we have left and then exit the loop, to unblock the `join`ed thread. 

Although, now that i think about it, it would be better to never join, should there be some heavy processing left to do. Is there a way to force GC on dead but not joined threads? Did I miss something and we don't need to join it at all?

Reply:
AH okay that makes sense, I'm with you. If we have some way of making the thread stop running (as we do now with the `break`), the `join` is unnecessary and probably would've been worse for the reasons you said WOOPS. Once the Thread completes, and nothing is referencing the dead thread, it'll get GC'ed. Or should.

Reply:
I'll remove the join then and we should be gtg, yeah?